### PR TITLE
Create a script to build the jupyter-web-app image and create a PR to update it

### PR DESCRIPTION
**Pull request by jlewi**
Create a script to auto build the Jupyter web app image and update the prototype
  
  * The script works as follows
  
    * Use git to determine the last commit at which the source to the Jupyter
      web app changed
  
    * Look for an image tagged with that commit
  
    * If no such image exists build a new image
  
    * Update the ksonnet prototype to use the new image
  
    * Push the commit to git; we will use the kubeflow-bot account
  
    * Use the hub CLI to create the pull request if one doesn't already exist
  
  * Create a Makefile to build the jupyter web app
  
    * Add a git label to the image so we can compare against the current image.
  
  * Put the new python code in python package kubeflow/kubeflow
  
    * We now have namespace packaging working
  
  * Provide a K8s job to run it. In a follow on PR we will turn this into
    a cron job
  
  Miscellaneous changes:
  
  Specify the build and publish projects separately.
  
  * Update pylintrcfile to always do no-self-use
  * Fix some lint issues

Related to #2206 - Auto update images

This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/3066)

### CODE REVIEW

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+  def update_prototype(self, image):
+    values = {"image": image}
+
+    regexps = {}
+    for param, value in values.items():
+      r = re.compile(r"([ \t]*" + param + ":+ ?\"?)[^\",]+(\"?,?)")
+      v = r"\g<1>" + value + r"\2"
+      regexps[param] = (r, v, value)
+
+    prototype_file = os.path.join(self._root_dir(),
```
Comment (richardsliu) [inline py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Should this be some sort of environment variable for the Kubeflow source root?

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+          replacements += 1
+          continue
+        prototype[i] = re.sub(regexps[param][0], regexps[param][1], line)
+        if line != prototype[i]:
+          replacements += 1
+    if replacements == 0:
+      raise Exception(
+          "No replacements made, are you sure you specified correct param?")
+    if replacements < len(regexps):
+      raise Warning("Made less replacements then number of params. Typo?")
```
Comment (richardsliu) [inline py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Why is the previous error an exception and this one just warning?

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    temp_file = prototype_file + ".tmp"
+    with open(temp_file, "w") as w:
+      w.write("\n".join(prototype))
+    os.rename(temp_file, prototype_file)
+    logging.info("Successfully made %d replacements", replacements)
```
Comment (richardsliu) [inline py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: What if replacement > len(regexps)? Could we accidentally replace something we're not supposed to?

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+    # TODO(jlewi):We should check what the current image and not update it
+    # if its the existing image
+    prototype_file = self.update_prototype(image)
+
+    branch_name = "update_jupyter_{0}".format(last_commit)
+
+    if repo.active_branch.name != branch_name:
+      logging.info("Creating branch %s", branch_name)
+      util.run(["git", "checkout", "-b", branch_name], cwd=self._root_dir())
```
Comment (richardsliu) [inline py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: The branch name could already exist if this script is run multiple times, for the same commit hash.

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    with open(prototype_file) as f:
+      prototype = f.read().split("\n")
+    replacements = 0
+    for i, line in enumerate(prototype):
+      for param, _ in regexps.items():
+        if param not in line:
+          continue
+        if line.startswith("//"):
+          prototype[i] = re.sub(
+            r"(// @\w+ )" + param + r"( \w+ )[^ ]+(.*)",  # noqa: W605
```
Comment (richardsliu) [inline py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Same here, maybe provide an example of what's being replaced.

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    # TODO(jlewi): Modeled on
+    # https://github.com/kubeflow/examples/blob/master/code_search/docker/ks/update_index.sh
+    # TODO(jlewi): We should use the GitHub API and check if there is an
+    # existing open pull request. Or potentially just use the hub CLI.
+
+    if not commit:
+      commit = self.last_commit
+      logging.info("No commit specified defaulting to %s", commit)
+
+    pr_title = "Update the jupyter-web-app image to {0}".format(commit)
```
Comment (richardsliu) [inline py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Add something like "Automated PR to update..." in the title.

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py (hunk 1 above)
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: The assumption is that you are running update_jupyter_web_app.py from the clone of the source repository that contains the source code you want to use. 
So we always build jupyter_web_app at the same version of the code we want to build

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py (hunk 6 above)
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Done

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py (hunk 5 above)
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: See above.

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py (hunk 4 above)
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Added a check.

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py (hunk 3 above)
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Got rid of this

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py (hunk 2 above)
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: Got rid of this

File: py/kubeflow/kubeflow/ci/update_jupyter_web_app.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+      data = yaml.load(hf)
+
+    return data["image"]
+
+  def update_prototype(self, image):
+    values = {"image": image}
+
+    regexps = {}
+    for param, value in values.items():
+      r = re.compile(r"([ \t]*" + param + ":+ ?\"?)[^\",]+(\"?,?)")
```
[the remark this answers is not in the document]
Comment (jlewi) [reply py/kubeflow/kubeflow/ci/update_jupyter_web_app.py]: I got rid of the regexes which are hard to read; and went with simple string parsing. Also added a unittest.

**Comment by jlewi** [comment]
@richardsliu This is ready for another look.

**Pull request merged**